cmd/reconciler/main.go

```diff
@@ -131,6 +159,122 @@ func main() {
+	// the GET reconcilerState on UPDATE/PATCH. So we need to use the non-caching client
+	// for the finalizer, which does GET/LIST after UPDATE/PATCH.
+	f := finalizer.New(reconcilerScope, supervisor, cl, // non-caching client
+		stopControllers, doneChannelForSyncer, doneChannelForRemediator)
+
+	// Create the Finalizer Controller
+	finalizerController := &finalizer.Controller{
+		SyncScope: reconcilerScope,
+		SyncName:  *syncName,
+		Client:    mgr.GetClient(), // caching client
+		Scheme:    mgr.GetScheme(),
+		Mapper:    mgr.GetRESTMapper(),
+		Finalizer: f,
+	}
+
+	// Register the Finalizer Controller
+	if err = finalizerController.SetupWithManager(mgr); err != nil {
+		klog.Fatalf("Instantiating Finalizer: %v", err)
+	}
+
+	// Configure Syncer and Remediator.
+
+	// Get a separate config for the remediator to talk to the apiserver since
+	// we want a longer REST config timeout for the remediator to avoid restarting
+	// idle watches too frequently.
+	cfgForWatch, err := restconfig.NewRestConfig(watch.RESTConfigTimeout)
+	if err != nil {
+		klog.Fatalf("Error creating rest config for the remediator: %v", err)
+	}
+
+	// Instantiate the cache shared by both the Syncer and Remediator
+	decls := &declared.Resources{}
+	remResources := cache.NewRemediateResources()
+
+	// Create the Remediator actor.
+	// The Syncer uses it to send pause/resume signals to the Remediator, and to start watches for declared resources.
+	// The Remediator uses it to receive the signal to perform remediation and to handle the remediate errors.
+	remInterface, err := remediator.New(reconcilerScope, *syncName, cfgForWatch, decls, remResources)
+	if err != nil {
+		klog.Fatalf("Error creating remediator interface: %v", err)
+	}
+	// reconcilerState is the cache shared across all Syncer reconciliation loops.
+	reconcilerState := parse.NewReconcilerState()
+	parser, err := configureParser(cl, discoveryClient, decls, supervisor, remInterface)
+	if err != nil {
+		klog.Fatalf("Error creating parser: %v", err)
+	}
+
+	// Create the Syncer controller
+	syncer := controllers.NewSyncer(
+		controllerCtx, reconcilerScope, parser, reconcilerState, *pollingPeriod, *resyncPeriod,
+		configsync.DefaultReconcilerRetryPeriod,
+		configsync.DefaultReconcilerSyncStatusUpdatePeriod, doneChannelForSyncer)
+
+	// Register the Syncer controller
+	if err = syncer.SetupWithManager(mgr); err != nil {
+		klog.Fatalf("Error creating controller %q: %v", controllers.SyncerController, err)
+	}
+
+	// Create the Remediator controller
+	rem := controllers.Remediator{
+		ControllerCtx:      controllerCtx,
+		DoneCh:             doneChannelForRemediator,
+		SyncScope:          reconcilerScope,
+		SyncName:           *syncName,
+		Applier:            baseApplier,
+		Resources:          decls,
+		RemediateResources: remResources,
+		Remediator:         remInterface,
+	}
+
+	// Register the Remediator controller
+	if err = rem.SetupWithManager(mgr); err != nil {
```

Comment:
When registering multiple controllers with the same manager, is there any guarantee that different controllers will not run concurrently?

The reason I ask is that we probably want to ensure the Syncer and Remediator controllers are never running at the same time

Reply:
Probably not. Kubernetes controllers operate independently, and their execution is not coordinated by default.

To ensure the Syncer and Remdiator controllers are not running at the same time, I used a shared flag `canRemediate`: https://github.com/GoogleContainerTools/kpt-config-sync/pull/974/files#diff-55e9dd0d45f70809718f2bafb46a7e8b0328caa9a2a018e442c3032ec35ba27cR46. Before the Syncer applies the declared resources, it pauses the remediator, and resumes it after the apply succeeds. The Remediator starts enqueueing and processing requests only when it the flag is true.

Reply:
Do you mean de-queuing and processing requests?

The current behavior is that watch events are still added to the queue, they just don't get handled until the applier exits successfully.

Reply:
The Remediator controller won't enqueue and dequeue+process requests if `canRemedite` is false.

Reply:
I made this comment a few patchsets ago, before synchronization was implemented for Syncer/Remediator. In the earlier patchset, it was possible for both the Remediator and Syncer to be running simultaneously.

The current patchset has synchronization which should prevent concurrent execution. The `CanRemediate` call doesn't prevent concurrent enqueueing given that's it's an unsafe boolean, but it is an optimization that should prevent some enqueueing of watch events. The actual synchronization is in the `StartRemediating` call in the reconcile loop

Reply:
Problem is, it needs to keep enqueuing requests, adding them to a queue and then handling them all when the applier succeeds. Otherwise it might miss drift that happened while the applier was running.